libs/ts-rest/nest/src/lib/api.decorator.ts

```diff
@@ -115,18 +118,57 @@ const getMethodDecorator = (appRoute: AppRoute) => {
 
 @Injectable()
 export class ApiRouteInterceptor implements NestInterceptor {
+  private isAppRouteResponse(
+    value: unknown
+  ): value is { status: HTTPStatusCode; body?: any } {
+    return (
+      value != null &&
+      typeof value === 'object' &&
+      'status' in value &&
+      typeof value.status === 'number'
+    );
+  }
+
   intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
     const res: Response = context.switchToHttp().getResponse();
 
+    const appRoute: AppRoute | undefined = Reflect.getMetadata(
+      tsRestAppRouteMetadataKey,
+      context.getHandler()
+    );
+
+    if (!appRoute) {
+      // this will respond with a 500 error without revealing this error message in the response body
+      throw new Error('Make sure your route is decorated with @Api()');
+    }
+
     return next.handle().pipe(
       map((value) => {
-        if (
-          typeof value === 'object' &&
-          typeof value.status === 'number' &&
-          value.body !== undefined
-        ) {
-          res.status(value.status);
-          return value.body;
+        if (this.isAppRouteResponse(value)) {
+          const statusNumber = value.status;
+
+          const isParsingEnabled = Boolean(
+            Reflect.getMetadata(ParseResponsesSymbol, context.getHandler()) ||
+              Reflect.getMetadata(ParseResponsesSymbol, context.getClass())
+          );
+
+          const responseValidation = checkZodSchema(
+            value.body,
+            isParsingEnabled ? appRoute.responses[statusNumber] : undefined
+          );
+
+          if (!responseValidation.success) {
+            const { error } = responseValidation;
+
+            const message = error.errors.map(
+              (error) => `${error.path.join('.')}: ${error.message}`
+            );
```

Comment:
Does this succeed if isParsingEnabled is false, leading the second param of checkZodSchema to be undefined? 

Alternatively, maybe we should early return if the validation isn't on 🤔 

Also, I'm aware this is a breaking change, primarily if someone is relying on receiving the Zod errors as an object.  On the flip side, the Zod error handling was undocumented behaviour, maybe we should discuss this change, then document + test it.

Reply:
@oliverbutler indeed that would be a breaking change, we consume the raw Zod errors in an exception filter, and rewrite them into our own error format.

@lukemorales can we separate this change into a separate issue to be discussed further?

Reply:
Sure thing, no problem! I thought of suggesting this approach for this as this would be a case were errors weren't being thrown for validation at the first place and perhaps a better formatted error message would be helpful to debug what's wrong with the parsed data. But it makes a lot of sense that people might be catching these errors and I will revert back to the original zod error

Reply:
>Does this succeed if isParsingEnabled is false, leading the second param of checkZodSchema to be undefined?
>
> Alternatively, maybe we should early return if the validation isn't on 🤔

It does success according to the `checkZodSchema` implementation, because the type guard accounts for undefined in it's implementation, there are even test cases written for that: https://github.com/ts-rest/ts-rest/blob/cb74bab9e26a8254adf4a05dc5e78c18b2249ff2/libs/ts-rest/core/src/lib/zod-utils.spec.ts#L41-L52

But just sharing for visibility, I agree with you and @michaelangrivera's comment that early return would be better for visibility